### DISCUSSION TITLE
Twig markdown_to_html filter usage: OK if MD is wrapped in filter's {%apply%} tags. "404 File not found" if try to include file and pipe to filter ?

### DISCUSSION THREAD

**Discussion by ghost** [symfony/symfony#42852] [Q&A] [answered]
Twig markdown_to_html filter usage: OK if MD is wrapped in filter's {%apply%} tags. "404 File not found" if try to include file and pipe to filter ?

I'm trying to include a markdown source file in my Twig template to then pipe to a Twig filter.

The Twig filter works fine if I don't pipe but just wrap the .md in Twig tags.

I'm doing something wrong when I use the include + pipe.  It never finds the source and I get a "404" error.

My twig templates are stored in my project dirs
```
project/
	resources/
		views/
			templates/
				pages/
					page1/
						page1.twig
						content.md
					page2/
						page2.twig
```

In other twig templates I can refer to those twig templates using template paths. like
```
{{ include('templates.pages.page1.page1') }}
```

I'm using the ```markdown_to_html``` filter (https://twig.symfony.com/doc/3.x/filters/markdown_to_html.html)

This renders markdown OK
```
cat page1.twig
	...
	<div>
		{% apply markdown_to_html %}
		# This is some markdown formatted text
		{% endapply %}
	</div>
	...
```

But if I try instead to include the source and pipe it to the filter,
```
cat content.md
	# This is some markdown formatted text

cat page1.twig
	...
	<div>
		{{ include('templates.pages.page1.content.md') | markdown_to_html }}
	</div>
	...

```

I just get a "404 Not Found"

I've also tried
```
cat page1.twig
	...
	<div>
		{{ include('content.md') | markdown_to_html }}
	</div>
	...
```

And
```
cat page1.twig
	...
	<div>
		{% apply markdown_to_html %}
			{{ include('content.md') }}
		{% endapply %}
	</div>
	...
```

All get the same 404 'Not Found' error.

The filter works in the wrapped-in-tags case so it looks like I've got the extension setup OK.

I'm betting it's something silly I've missed, but so far I'm stumped.

How do I get the md in the file included into my twig template and used by the twig filter?

I'd be grateful for any help!

Thanks!

**Comment by stof** [comment]
I don't get what you say with `I just get a "404 Not Found"`. Twig is not responsible for returning HTTP responses at all.

**Comment by ghost** [comment]
I'm of course talking about what I see in the browser after compiling the Twig templates to rendered HTML.
If I add the md code directly to the template, wrapped in those {%apply%} tags, then it renders fine -- I see the md in the browser.
If I do it with the 'include', I get the 404.  The source isn't found.

**Reply by stof** [reply]
Twig will not be able to generate a HTTP response (404 or other btw). So if you get a 404, something else (not described in your question) is involved.

**Comment by ghost** [comment]
I didn't SAY that Twig generates the response.

I said: "I'm betting it's something silly I've missed, but so far I'm stumped."

The ONLY thing I change to get a 404 Error showing up in my site is the 'md file include.  And that's ONLY used with this filter.

Other filters work fine.
Other includes work fine.
Markdown works fine if I just wrap the code like I said.

If I include the file it's not found.

I don't know WHAT the problem is.
I know that the way to 100% REPRODUCE the problem is to use the include in my Twig file.

**Reply by stof** [reply]
> I don't know WHAT the problem is.
> I know that the way to 100% REPRODUCE the problem is to use the include in my Twig file.

That's not a full reproducer, given that Twig does not generate HTTP responses. Something else in your stack will do it, which  is not described in your question.

**Comment by ghost** [comment, marked as answer]
> That's not a full reproducer,

Uh-huh.
never mind.

**update**
turns out that it IS a Twig issue, not "something else in your stack".

The docs

https://twig.symfony.com/doc/3.x/filters/markdown_to_html.html

say
```
You can also use the filter on an included file or a variable:

{{ include('some_template.markdown.twig')|markdown_to_html }}
```

If you use either of
```
some_template.markdown.twig
some_template.markdown
```

you get the 404 when you look at your page in a browser.

If you get rid of the first period separator
```
-	some_template.markdown.twig
+	some_template_markdown.twig
```

and remove the .twig extension, then as long as there's only one extension dot in the name, then using```
{{ include('your/path/to/some_template_md')|markdown_to_html }}
```

works.

I.e.
the content filename needs to end in .twig
the content filename can have only one dot
the include should NOT use the .twig extension

now, works great & everywhere
